lib/signs/utilities/messages.ex

```diff
@@ -23,252 +26,287 @@ defmodule Signs.Utilities.Messages do
-
-          {Content.Message.Custom.new(line1, :top), Content.Message.Custom.new(line2, :bottom)}
-
-        sign_config == :off ->
-          {Content.Message.Empty.new(), Content.Message.Empty.new()}
-
-        sign_config == :headway ->
-          get_alert_messages(alert_status, sign) ||
-            Signs.Utilities.Headways.get_messages(sign, current_time)
-
-        true ->
-          case Signs.Utilities.Predictions.get_messages(predictions, sign) do
-            {%Content.Message.Empty{}, %Content.Message.Empty{}} ->
-              if sign_config == :temporary_terminal and
-                   alert_status in [:shuttles_transfer_station, :suspension_transfer_station] do
-                Signs.Utilities.Headways.get_messages(sign, current_time)
-              else
-                get_alert_messages(alert_status, sign) ||
-                  Signs.Utilities.Headways.get_messages(sign, current_time)
-              end
-
-            {top_message, %Content.Message.Empty{}} ->
-              {top_message,
-               get_paging_headway_or_alert_messages(sign, current_time, alert_status, :bottom)}
-
-            {%Content.Message.Empty{}, bottom_message} ->
-              if match?(
-                   %Content.Message.Predictions{station_code: "RJFK", zone: "m"},
-                   bottom_message
-                 ) do
-                jfk_umass_headway_paging(bottom_message, sign, current_time, alert_status)
-              else
-                {get_paging_headway_or_alert_messages(sign, current_time, alert_status, :top),
-                 bottom_message}
-              end
-
-            messages ->
-              messages
-          end
-      end
+    cond do
+      match?({:static_text, {_, _}}, sign_config) ->
+        {:static_text, {line1, line2}} = sign_config
+
+        {Content.Message.Custom.new(line1, :top), Content.Message.Custom.new(line2, :bottom)}
+
+      sign_config == :off ->
+        {%Content.Message.Empty{}, %Content.Message.Empty{}}
+
+      match?(%{source_config: {_, _}}, sign) ->
+        Enum.zip([
+          Tuple.to_list(sign.source_config),
+          Tuple.to_list(predictions),
+          Tuple.to_list(alert_status),
+          Tuple.to_list(scheduled),
+          Tuple.to_list(service_status)
+        ])
+        |> Enum.map(fn {config, predictions, alert_status, scheduled, service_status} ->
+          predictions =
+            filter_predictions(predictions, config, sign_config, current_time, scheduled)
+
+          alert_status = filter_alert_status(alert_status, sign_config)
+
+          Signs.Utilities.Predictions.prediction_message(predictions, config, sign) ||
+            service_ended_message(service_status, config) ||
+            alert_message(alert_status, sign, config) ||
+            Signs.Utilities.Headways.headway_message(sign, config, current_time) ||
+            early_am_message(current_time, scheduled, config) ||
+            %Content.Message.Empty{}
```

Comment:
Wow this is excellent and so much easier to understand the intent compared to the old code. Cool to see how far the vision has come!